### ISSUE TITLE
[Question] Consider migration from GitHub

### ISSUE THREAD

**Issue by FireMasterK** [blackjack4494/yt-dlc#9]
## Checklist

- [x] I'm asking a question
- [x] I've looked through the README and FAQ for similar questions
- [x] I've searched the bugtracker for similar questions including closed ones

## Question

Have you considered migrating from GitHub to prevent this entire situation from happening again?

A mirror is now available at https://git.kavin.rocks/kavin/yt-dlc
Tweet regarding the CEO trying to bring back youtube-dl: https://nitter.kavin.rocks/t3rr4dice/status/1320660235363749888

**Comment by ssokolow** [comment]
If so, I'd suggest a European-hosted alternative, like the [Gitea](https://gitea.io/) instance at [Codeberg.org](https://codeberg.org/).

**Comment by xloem** [comment]
https://git-annex.branchable.com/ uses a solution where issues etc are stored in the same git repository as the source code.

**Comment by Mr-Steal-Your-Script** [comment]
just here to say fuck youtube

**Comment by Redo11** [comment]
The fact that this is still here is amazing. I really want this source code to be backed somewhere safe, but be aware that self-hosted options place you at the front of dealing with DMCA. The safest way would be to use something so obscure, that they ain't even bother with claiming it.

**Comment by Mr-Steal-Your-Script** [comment]
someone should make a simple html file that downloads a zip file of the repo

**Comment by ssokolow** [comment]
@Redo11 Aside from Codeberg, here are few more options I've since learned of:

* [Gitee](https://gitee.com/) is based in China.
* [Git-SSB](https://git.scuttlebot.io/%25n92DiQh7ietE%2BR%2BX%2FI403LQoyf2DtR3WQfCkDKlheQU%3D.sha256) is a decentralized solution which self-hosts its own development on the Secure Scuttlebutt overlay network.
* [CodeFuse](https://codefuse.eth.link/#!/settings?t=about) is apparently a GitHub clone sitting on top of [WebIndexP2P](https://wip2p.eth.link/), but I haven't yet figured out if it's open-source.

Also, [SourceForge](https://sourceforge.net/) is so desperate to stay relevant that they welcome people using SourceForge to set up supplementary services for projects hosted elsewhere, which might help with DMCAs since the code and the non-code hosting would be independent. (For the common case of projects on GitHub, they even have a *wizard* for creating a SourceForge project to supplement a GitHub repo.)

**Comment by xloem** [comment]
@ssokolow @Redo11

more things, not in a particular order

- decentragit  https://github.com/quorumcontrol/dgit
- gittorrent https://github.com/cjb/GitTorrent
- git-remote-bsv https://github.com/xloem/git-remote-bsv
- ganarchy #11
- git-remote-ipfs https://github.com/dhappy/git-remote-ipfs
- ikiwiki code management as i mentioned earlier, example at https://git-annex.branchable.com/

**Comment by Seirdy** [comment]
a simpler alternative would be de-coupling the issue-tracking from the repo. several mirrors could be put in place (github, gitlab, sourcehut, bitbucket, codeberg, notabug.io, repo.or.cz, a bunch of self-hosted instances) so you get a ["hydra" effect](https://www.youtube.com/watch?v=qlwVTRTCdnk). Git lets you work with multiple remotes out of the box.
Then the issue becomes figuring out how to keep issue-tracking resilient. The best solution would be mailing lists; since everybody has a copy of everything in their maildir, the mailing list is fully distributed/decentralized just like git. The RIAA is less likely to come for a mail server, and if they do, migration is trivial if you have a maildir or MBOX.

**Comment by ssokolow** [comment]
@Seirdy Dear *God* please no. I *despise* when circumstances force me to join a mailing list... unless you're willing to guarantee the availability of an NNTP bridge.

**Comment by Seirdy** [comment]
@ssokolow the thing is, most mainstream issue trackers require you to create an account. That in itself is "too onerous" for people who already have an email associated with their git identity. Asking everyone to create a codeberg/gitlab/notabug/sf account probably won't fly.

If you consider emailing a patch too onerous, you could just post a pastebin link to IRC instead. There's no shortage of ways to send a file that don't revolve around putting all your eggs in the same basket that's hosting your repo.

My mailing-list suggestion was predicated on the idea that switching a git remote is extremely trivial, while migrating issues/patches is more tricky. Email is already a proven solution for some of the largest git projects in existence.

If you have another well-tested solution for de-coupling issue tracking from the code forge and distributing it among participants (so everyone gets an offline copy) without requiring contributers to create yet another account, I'm all ears.

**Comment by ssokolow** [comment]
@Seirdy As long as there are separate mechanisms for contributors and mere bug reporters, and bug reporters don't have to do something like registering a throaway account to keep the mailing list from cluttering up their inbox, I'll be happy with it.

That said, I do remember running across a couple of experiments in retrofitting Git with Fossil-style "bug tracker data stored in the repo" issue trackers. I wonder if any of those are still active.

**Comment by Seirdy** [comment]
> bug reporters don't have to do something like registering a throway
> account to keep the mailing list from cluttering up their inbox

Having to create an account somewhere, be it on GitHub or
elsewhere, typically requires entering an email address. By default,
replies in issues get sent to your email, just like a mailing list.
Either way, you're signing up for emails and it's up to you to then
control whether or not they clutter up your inbox.

Something like Sourcehut offers a nice web UI for mailing lists to make
things easier for people who don't use email.

Learning to use git-bug and building a stable web UI for it sounds like
overkill when it's probably easier to just send an email.

Git+email is already decentralized, widely used, and stable. People
already know how to email attachments. Being able to use email is a 
re-usable skill that can then be applied to other mailing-list-based 
projects, reducing lock-in to companies like MS.

Let's not re-invent the wheel to introduce a new, more complex tool for
everyone to learn just to contribute to youtube-dl. If you really want a
web UI, Sourcehut already has one that lets you pretend you aren't
using a mailing list.

**Comment by ssokolow** [comment]
As long as I'm not being forced to reinvent "just one thread and mentions elsewhere, please" filtering and have a simple unsubscribe link/button that doesn't require me to send an UNSUBSCRIBE e-mail that's formatted just right and/or appears to comes from just the right inbound-only alias in my "spam defense via e-mail addresses as revocable API keys" mail system, I'm fine.

**Comment by xloem** [comment]
> **UPDATE:** I think [Bugs Everywhere](https://gitlab.com/bugseverywhere/bugseverywhere) was one of the ones I originally remembered.

A first glance at this solution appears to make sense here.  It's written in python, just like this project, so contributors would be at home. _It already has a web interface and an email interface_.

I don't know why you'd want to decouple the bugs from the code when the project may get disrupted.  Combining them helps preservation and access.

One concern was learning to use these things.  Somebody has to put in the effort of setting it up.

Some other suggestions were for tools that bridge between github, gitlab, jira.  These have the advantage of not requiring anyone to run a server to provide easy user access.

**Comment by BiosPlus** [comment]
I'd recommend notabug.org

Would also recommend looking into setting up a mirror of the primary repo on a self hosted gitea or gitlab instance hidden behind a domain from njal.la

**Comment by rain-1** [comment]
> a simpler alternative would be de-coupling the issue-tracking from the repo. several mirrors could be put in place (github, gitlab, sourcehut, bitbucket, codeberg, notabug.io, repo.or.cz, a bunch of self-hosted instances) so you get a ["hydra" effect](https://www.youtube.com/watch?v=qlwVTRTCdnk). Git lets you work with multiple remotes out of the box.
> 
> Then the issue becomes figuring out how to keep issue-tracking resilient. The best solution would be mailing lists; since everybody has a copy of everything in their maildir, the mailing list is fully distributed/decentralized just like git. The RIAA is less likely to come for a mail server, and if they do, migration is trivial if you have a maildir or MBOX.

Seconded, How about https://github.com/MichaelMure/git-bug

**Comment by vxbinaca** [comment]
Just do Gitee and call it a day, this magic shit about IPFS and decentralized stuff is overkill. The Chinese don't care.

**Comment by xloem** [comment]
If anyone can actually set any of these up so others can use them, that's what the community really needs here.

**Comment by jbruchon** [comment]
Removal of the references to downloading copyrighted material is sufficient. The DMCA takedown's only realistic leg to stand on was the test cases showing that the primary intent of this program was to be used to obtain copyrighted material. Clearly, this is a tool that can be used for good or evil, and archivists like myself use it to back up YouTube channels that could be at risk of being lost in the future as YouTube continues its slow shift towards becoming "cable TV, but online." There is no longer any solid footing that a new DMCA takedown notice can stand on. The program doesn't circumvent copyright protection mechanisms (the JavaScript to assemble media stream URLs and the media streams themselves are all sent by YouTube unencrypted without a browse-wrap license agreement and/or mandatory user registration, and no, SSL is transport layer encryption, not DRM) so it now falls into the same category as a tool like HandBrake: it CAN be used to do things that are possibly a violation of copyright, but that is neither its only purpose nor its primary purpose.

If you move from GitHub, you're moving off of the largest open source software hosting platform on the planet. You will lose searchability and a lot of people simply won't register at the other lesser-known sites to contribute to the code base.

I don't think moving is a good idea.

**Comment by Seirdy** [comment]
If you want to maintain the discoverability and searchability, just make 
GitHub one of your remotes. It doesn't have to be the primary remote. 
This goes back to the "Hydra effect" I mentioned earlier.

The ideal solution would be to have a plethora of remotes but one 
bug-tracker. Just `git remote add` GitHub on top of whatever else you 
want to use.

I'm not in favor of bundling the bug-tracker with the repo; getting the 
code should not require getting all the issues and patches, and simple
tasks like responding to an issue with instructions to reproduce should
not require users to create an account, clone the repo and use git; that
would impose a more significant barrier to entry than sending an email.

I feel like a broken record when I say that git is already distributed,
federated, and decentralized. A plethora of projects much more
well-known than youtube-dl leverage this. We shouldn't require people to
learn a brand-new tool and make an account just to contribute to
youtube-dl when we can be a good FOSS citizen and encourage people to
use a process that can be re-used in other projects.

Much of youtube-dl's functionality leverages ffmpeg, which also uses 
mailing lists and a GitHub mirror.

**Comment by 322997am** [comment]
I believe the best way is to self-host on a VPS service in a country that ignores the DMCA(Russia for example). The DMCA is a bad law that is probably unconstitutional, but stuff like these takedowns will continue happening as long as Disney and the like exist. I believe that Gitea is feature rich and can be self-hosted. I am willing to help with any translation needed if this is ever decided on, as I am fluent in both Russian and English.

**Comment by ssokolow** [comment]
@322997am ...but do mirror everything somewhere else or your [bus factor](https://en.wikipedia.org/wiki/Bus_factor) is one.

**Comment by xloem** [comment]
If somebody sets up a mirror, can anybody volunteer to co-maintain or co-administer it?

**Comment by FireMasterK** [comment]
https://git.kavin.rocks/kavin/yt-dlc

Here's a mirror on my personal git server.

**Comment by FireMasterK** [comment]
A relevant tweet to this topic: https://twitter.com/t3rr4dice/status/1320660235363749888

**Comment by Seirdy** [comment]
Just thought I’d share my approach to “hydra hosting” since people here seem interested:

I mirror my repos across Sourcehut, Gitlab, and GitHub. Here’s the relevant snippet of my `.git/config` of my dotfiles repo:

``` gitconfig
[remote "origin"]
	url = git@git.sr.ht:~seirdy/dotfiles
	fetch = +refs/heads/*:refs/remotes/origin/*
[remote "gl_mirror"]
	url = git@gitlab.com:Seirdy/dotfiles.git
	fetch = +refs/heads/*:refs/remotes/gl_mirror/*
[remote "gh_mirror"]
	url = git@github.com:Seirdy/dotfiles.git
	fetch = +refs/heads/*:refs/remotes/gh_mirror/*
```

Pushing to three remotes, one after the other, can be slow. To speed things up, I created an alias to push to all remotes in parallel in [my gitconfig](https://git.sr.ht/~seirdy/dotfiles/tree/master/.config/git/config):

```  gitconfig
[alias]
	pushall = !git remote | grep -E 'origin|mirror' | xargs -L1 -P 0 git push --all --follow-tags
```

This pushes to all remotes that have "origin" or "mirror" in their names, and skips the rest. Now, I can pull from all repos I'm following and push to the ones I have access to.

Core developers can post a list of a few upstream remotes that only they can push to. Community members can set up extra remotes for resiliency and for forking/personal development.

This raises an issue: when there are multiple git remotes, where do people file tickets or submit patches?

The best solution is to have one canonical place to do issue tracking, separate from git remotes. I've previously explained why I think a mailing list (with a Sourcehut-style frontend for those not used to mailing lists) would be the best option for this, but a plethora of other solutions exist as well, from GitHub issue trackers to Bugzilla.

**Comment by FireMasterK** [comment]
Why not have a cronjob to do the same? This way even actions on github could be added and the developers don't need to change their work flow / modify their git configs when pushing commits

**Comment by ssokolow** [comment]
@FireMasterK What I've been meaning to set up for my own projects is:

* Use something like a cronjob or a hook to automatically merge updates on my GitHub issue trackers into the repo using something like [git-bug](https://github.com/MichaelMure/git-bug) so my users get the interface they're used to, but the issues get merged into whatever backup/mirroring regime I choose.
* Set up automatic mirroring between GitHub, GitLab, and BitBucket (either using ready-made support like [Gitlab offers](https://docs.gitlab.com/ee/user/project/repository/repository_mirroring.html) or by adapting the code I use for rebuilding `gh-pages` on `git push` to drive the syncing.

**Comment by Seirdy** [comment]
@FireMasterK The whole point of this approach is not to depend on one service, be it a git remote or a CI/CD provider, especially a proprietary one, like GitHub Actions. These features have been baked into git and successfully used for a long time.

**Comment by ddevault** [comment]
SourceHut admin here. I wrote this up today, is relevant:

https://sourcehut.org/blog/2020-10-29-how-mailing-lists-prevent-censorship/

**Comment by ssokolow** [comment]
@ddevault I fully agree, ideologically... I'm just already too overwhelmed with other things in my life to justify a UI/UX regression in the name of principles, and I doubt I'm the only one.

**Comment by ddevault** [comment]
Many people, myself included, do not believe sourcehut is a UI/UX regression. I hear positive feedback on the UI all the time, and by objective measures, it is the [fastest and most accessible forge](https://forgeperf.org).

**Comment by ssokolow** [comment]
@ddevault Does it provide features like:

* Receiving e-mails only for specific threads
* Receiving e-mail notification of someone using an `@mention`?
* Subscription/unsubscription control via HTTP, rather than an unreliably particular e-mail interface involving sending certain messages from certain addresses?

If so, then it's not what I think when I hear "mailing list" but more "forum software with best-in-class e-mail integration".

**Comment by ddevault** [comment]
>Receiving e-mails only for specific threads

No, but this is planned and prioritized.

>Receiving e-mail notification of someone using an @mention?

Not on the mailing lists. Just Cc them. The bugtracker does support ~mentions though.

>Subscription/unsubscription control via HTTP, rather than an unreliably particular e-mail interface involving sending certain messages from certain addresses?

Yes.

**Comment by ssokolow** [comment]
> No, but this is planned and prioritized.

> Yes.

Those are my two biggest issues taken care of them.

> Not on the mailing lists. Just Cc them.

Generally, as an anti-spam measure, I try to configure my mail system so that any messages not coming from an authorized sender (in this case, the listserv address) will get redirected to trash without me seeing them.

(When I have time to implement a custom milter, it'll be fully automatic, where outbound mail to someone generates a custom `From:` address, a browser extension lets me generate one for a form from the context menu, and any attempt by... say... an eBay seller to subscribe my PayPal address to their newsletter will just result in "Bounced: You tried to send a message to my PayPal address but you are not PayPal.")

**Comment by ddevault** [comment]
>Generally, as an anti-spam measure, I try to configure my mail system so that any messages not coming from an authorized sender (in this case, the listserv address) will get redirected to trash without me seeing them.

Our mailing lists (unlike some others, I know, I know, I hate it too) will not break DKIM signatures, which is a better measure of message authenticity than checking the sender.

**Comment by ssokolow** [comment]
DKIM doesn't prevent a "legitimate" sender from sending unsolicited e-mail to an address that was shared with them without my permission.

Thanks to my manual use of tools like SpamGourmet, all the spam I can remember getting in the last decade has been something where things like SPF and DKIM are set up correctly... the sender just isn't the person I gave that alias to.

It gets annoying having to change things like my "from eBay" address just because eBay leaked my e-mail to a Chinese seller I bought something from once, and China doesn't have the kind of strong penalties for unsolicited commercial mail that [we have](http://fightspam.gc.ca/) here in Canada.

**Comment by ddevault** [comment]
>Thanks to my manual use of tools like SpamGourmet, all the spam I can remember getting in the last decade has been something where things like SPF and DKIM are set up correctly... the sender just isn't the person I gave that alias to.

You could also subscribe to the mailing list using an alias address and check the rcpt to.

**Comment by ssokolow** [comment]
> You could also subscribe to the mailing list using an alias address

That's a given. My inbox is configured to trash any message that didn't arrive via an alias.

> and check the rcpt to.

I try to avoid implementing custom fixes on a per-site basis.

**Comment by ddevault** [comment]
*shrug*

This is how mailing lists work. Your setup is exceptional, not the norm, and at some point the chickens come home to roost. These emails are legitimate and we've taken many steps to make sure that they are authenticated as such.

**Comment by ssokolow** [comment]
Hey, as long as people don't expect to be able to CC me without first requesting to be added to my whitelist and have the message arrive, I'm fine with status quo on that point.

**Comment by xloem** [comment]
If bug discussion threads were inside git repositories, we could add a commit to move the long conversation between 2 people away from the main thread.

**Comment by FrickTheRIAA** [comment]
I have put together a guide on how to host yt-dlc offshore using companies that have a track record of being very resilient to DMCA takedowns: https://old.reddit.com/r/youtubedl/comments/jktx5b/how_to_anonymously_host_the_continued_development/
I won't copy paste the whole guide into this post since 1) it might be better to just link to it off-site and have further away from the development of the software and 2) it's pretty long.

**Comment by xloem** [comment]
> https://git.kavin.rocks/kavin/yt-dlc
> 
> Here's a mirror on my personal git server.

@FireMasterK are you available at all to run a cron job or service on your host that would run something like [git-bug](https://github.com/MichaelMure/git-bug)?

**Comment by BiosPlus** [comment]
imo this should be marked as closed.

- This repo is already being backed up by the maintainer in a separate self-hosted location.
- The current CEO of GH is working with the YTDL team to have the original repo reinstated (he's present on their IRC).
- This action exists that can be used to mirror the repo elsewhere. (https://github.com/spyoungtech/mirror-action)

**Comment by xloem** [comment]
> * This repo is already being backed up by the maintainer in a separate self-hosted location.

obviously that doesn't address the bus-factor concern, but do you have a reference to where to find that mirror please?

**Comment by BiosPlus** [comment]
> obviously that doesn't address the bus-factor concern, but do you have a reference to where to find that mirror please?
It's being kept private from what I can see, the maintainer is on a discord that we share and has posted it. I'll leave it to them to choice whether to divulge the url

**Comment by shakeyourbunny** [comment]
>>   https://git.kavin.rocks/kavin/yt-dlc
>>  Here's a mirror on my personal git server.
>@FireMasterK are you available at all to run a cron job or service on your host that would run something like git-bug?

This site uses gitea, which has an inbuilt feature to mirror other repos.

**Comment by xloem** [comment]
@shakeyourbunny do you know whether gitea mirror issues like git-bug does, and pull requests?

**Comment by Nekun** [comment]
@BiosPlus
> The current CEO of GH is working with the YTDL team to have the original repo reinstated (he's present on their IRC).

https://nitter.net/t3rr4dice/status/1320660235363749888
As you can see on screenshot, one of conditions for return youtube-dl repository on GitHub is deleting code to bypass YouTube DRM, so if youtube-dl team will make decision to stay on GH, functionality will be strongly reduced, you'll can't download or watch in mpv much of music videos anymore.

**Comment by xloem** [comment]
there also doesn't appear to be a discord channel.  there's a gitter bridged to matrix.

**Comment by Nekun** [comment]
> I have put together a guide on how to host yt-dlc offshore using companies that have a track record of being very resilient to DMCA takedowns: https://old.reddit.com/r/youtubedl/comments/jktx5b/how_to_anonymously_host_the_continued_development/

This post was deleted from Reddit, fortunately I had saved a copy, very good guide:
https://nekun.cpunks.moe/ul/how-to-anonhost.html

**Comment by FrickTheRIAA** [comment]
@Nekun Thanks for the kind words. Yeah, it seems like it was removed. I left a comment about it in the thread, but that comment was instantly hidden. Feels a bit shitty, I put more time into this than I should have. I haven't broken any rules as far as I'm aware and nothing mentioned in my post is illegal. I basically wrote an almost 4000 word well-researched essay with citations and everything, so a sentence in return from whoever removed my thread (be it mods or Reddit themselves) doesn't seem like too much to ask.

Anyway, I've reuploaded it to Disroot's CryptPad. Good luck taking it down from there.
This is an updated version, the most notable change was that I noticed that NiceVPS allows for completely anonymous domain registration, just like Njalla. If you could update your mirror to reflect the changes, that would be great. Thanks!
[Viewing mode](https://cryptpad.disroot.org/code/#/2/code/view/GcymuR0ll9qlGbYtCvI89Q6WxKTT0wvmboh4lnNt4rA/embed/present/) (text only)
[Markdown mode](https://cryptpad.disroot.org/code/#/2/code/view/GcymuR0ll9qlGbYtCvI89Q6WxKTT0wvmboh4lnNt4rA/embed/) (in case you want to save the original document with links and everything)

**Comment by vxbinaca** [comment]
RIAA just wants to disrupt it in a way thats easy for them, they won't go after gitee in china, or any other foreign country. 

What matters right now is getting up and running. Anything else is wasted energy rn. Youtube-dls codebase is rotting and was before they were removed. I do not count on the devs being assed to defend a codebase that had 800 PRs in que.

**Comment by xloem** [comment]
if you don't want your files to disappear, using things that use hash-indexing can help ( https://ipfs.io/ , https://siasky.net/ ), and a blockchain is pretty much impervious once you get it up ( https://bico.media/ )

**Comment by ohnonot** [comment]
The code will remain, that much is certain.

What bothers me is the much more fragile dynamic infrastructure of communication, issues etc. 
I believe it is possible to have that mirrored as well (or most of it) with some git sites?
Not only a static mirror, but constantly parallel on the same level as this repo here.
Does this already exist, or not? I'm sorry if I could not extract this information from all those comments...

The repo owner has not commented on this?

In the long run I think one should strive to move all development communication to that other site.

**Comment by xloem** [comment]
@FireMasterK how would you feel about editing your first post so as to include some of the information in the thread for new people?

@ohnonot there are a lot of proposed decentralized solutions but not much evidence of anybody putting work into setting them up to use.  I'm working a little on migrating archived issues to https://github.com/MichaelMure/git-bug in the #youtube-dl freenode irc channel.  I don't know if I will finish, but a couple have put some community work into doing something about the issues, prs, communication channels.  secure scuttlebutt is a nice technology.  so is pgp-signed email.  There's been a lot of discussion on setting up git mirrors that automatically synchronize.  If you're running one, please link it to any others you can find.

The main youtube-dl repo owner is currently tied up in proceedings with EFF and github's ceo, and has said they can't communicate about the details.

ther'es a gitea instance set up inside onionland ( needs the browser bundle from https://torproject.org/ ) at githidep2hynhdmutuv7n2tei4iie2c7lyqz5fes3r5zzoxe5dshtxyd.onion and some sources listed at https://ipfs.io/ipns/QmbgyxFYSB29zFCnLYVQANjZghJw2cCAGpczLojV3g39yY/

**Comment by ohnonot** [comment]
> @ohnonot there are a lot of proposed decentralized solutions but not much evidence of anybody putting work into setting them up to use.

It's not hard to set a repository up to be in sync with a github repo; I propose framagit or any other gitlab-based git site: https://docs.gitlab.com/ee/user/project/repository/repository_mirroring.html (not sure how to make that go both ways though)
the most important thing is to leave github behind IMO.

> The main youtube-dl repo owner is currently tied up in proceedings with EFF and github's ceo, and has said they can't communicate about the details. i'm not in touch myself with blackjack4494, the owner of this community fork.

Thanks for that additional info. Appreciated.

I'm not sure it's necessary to go TOR on this; IMO just moving out of the RIAA's reach is enough.

**Comment by Nekun** [comment]
@vxbinaca 
>Anything else is wasted energy rn.

It's naive to think that copyright holders will immediately gives up after simple counter-measures and will not make troubles for project in near future. Just look on Popcorn Time app, how many domains they already changed? Why don't be on more than one step forward in this whack-and-mole game and improve resistance standard for that kind of projects?

**Comment by SoniEx2** [comment]
You could use git-bug to setup a decentralized/federated GAnarchy-based issue tracker, fwiw. While also doing decentralized development on GAnarchy. It's kinda your best bet to continue development while not excluding any potential contributors - setting up a centralized git instance (such as gitlab or gitea) on tor or i2p is a good way to exclude a lot of contributors, when there's no reason said contributors can't temporarily push their fork on github/gitlab.com/bitbucket/etc despite the risk of their repo getting taken down, so that other developers can quickly grab a copy and merge it.

**Comment by Seirdy** [comment]
On Sat Oct 31, 2020 at 6:44 AM PDT, ohnonot wrote:

> It's not hard to set a repository up to be in sync with a github repo; I
> propose framagit or any other gitlab-based git site:
> https://docs.gitlab.com/ee/user/project/repository/repository_mirroring.html
> (not sure how to make that go both ways though)

How is this any better than just adding a remote the way I described? 
That way, you aren't locked into GitLab's API and you don't have to 
learn any more tools (it's the same as adding the origin remote).

> I'm not sure it's necessary to go TOR on this; IMO just moving out of
> the RIAA's reach is enough.

Adding a remote for TOR doesn't hurt. Nobody is stopping you from 
setting up a mirror. It can only help.

**Comment by ssokolow** [comment]
@Seirdy All the methods I've seen for using git remotes involve either client-side intervention or something like a server-side cronjob.

Gitlab's mirroring support is a much more setup-and-forget solution and has support for event-driven configurations on the server side.

**Comment by xloem** [comment]
to clarify, it sounds like users are interested in helping but aren't sure how to run a mirror.  i heard gitolite can help with this, i haven't looked it up.  there are a lot of options, easiest of which are places like codeberg and sourcehut.  what's important is to start a norm of listing the mirrors in the repo; we can write scripts to sync them after that

**Comment by SoniEx2** [comment]
you can also use plain HTTPS to host git repos.

**Comment by virtadpt** [comment]
Some of us use [Fossil](https://fossil-scm.org/) to avoid these problems.  But use whatever people are willing to use.

**Comment by blackjack4494** [comment]
I haven't forgotten this topic if some may wonder.

**Comment by xloem** [comment]
@blackjack4494 i've been putting some work into learning some of the internals of git-bug while playing around with making a python hack to import issues from sausage's gharchive dumps ( https://github.com/xloem/youtube-dl-1/blob/gitbugideas/devscripts/sausage2gitbug.py i'm working on that roughly constantly right now, but i don't think it's the most efficient approach ).  curious if any other projects or plans are in the works.

**Comment by Redo11** [comment]
Coming back, @xloem and @ssokolow, I know one more good service to host git repositories, encrypted etc, called Keybase. It's kinda like slack+LinkedIn+git, but is encrypted and everything stored there is safe. That would mean, we can communicate and work on code etc, but one problem is that it is also private. Next solution would be IPFS, which is P2P file hosting(Idk, how to better describe it, check it out), it is great, but also pros and cons should also be considered. Next option is going onto dark net like ZeroNet or I2P (tor and FreeNet are too slow).

**Comment by FrickTheRIAA** [comment]
Keybase is great. I'm just unsure how long it will last since Zoom bought it to get encryption specialists to fix their security mess and save their reputation. I wouldn't be surprised if Keybase gets shut down eventually with all of the employees working on Zoom full time. Then there's the anti-end-to-end encryption legislation that the US government is pushing that might impact Keybase, though the EU is apparently working on anti-end-to-end encryption legislation as well, so the free world is shrinking. Keybase is great, just don't get shocked if you have to migrate from there within a few years.

**Comment by FrickTheRIAA** [comment]
I just noticed that [Disroot has a Gitea instance](https://git.disroot.org/). While I still firmly believe that self-hosting with a DMCA ignored hosting provider is better, Disroot's Gitea instance would be one of the best already existing platforms for a project like this. It's not immune to to DMCA, but based on the links below I have a feeling that both Disroot and Greenhost (Disroot's hosting provider) would be as upset as we are if they got a DMCA takedown order for an open source project and that they would probably try to fight it. Extra plus to Disroot for being Copyleft and having a good [privacy policy](https://disroot.org/en/privacy_policy).
[Disroot's About page](https://disroot.org/en/about)
[Disroot's Mission Statement](https://disroot.org/en/mission-statement)
[Greenhost - Internet Freedom](https://greenhost.net/internet-freedom/)
[Privacy at Greenhost](https://greenhost.net/privacy/)

**Comment by FrickTheRIAA** [comment]
@Nekun I made a big mistake in my guide regarding Tor over VPN, it's not a good idea at all. Can you please update your mirror one last time? I've added links back to the original Reddit and Raddle threads into the guide now, so you wouldn't have to update your mirror after this thanks to that.
(Sorry for this off-topic post, but I don't know how else to contact Nekun. This will be the last post like this.)

By the way, [a mod managed to restore the original Reddit thread](https://old.reddit.com/r/youtubedl/comments/jlnnzg/backup_of_how_to_anonymously_host_the_continued/gaqyinj/) and [confirmed that the thread was not removed by the /r/youtubedl mods](https://old.reddit.com/r/youtubedl/comments/jktx5b/how_to_anonymously_host_the_continued_development/#gao9ff2), which means that it was indeed removed by Reddit.

**Comment by xloem** [comment]
@blackjack4494 , I'm not sure how to get in contact with you.  The following is from the guy who runs https://youtube-dl-sources.org/

> Does youtube-dlc have any sort of DNS presence? youtube-dlc.org seems to
> be unregistered. e.g. https://bugs.youtube-dlc.org could be pointed to a
> hosted Gitea instance, as you mention, with some links to mirrors,
> places to report bugs, etc. I'd suggest that Tom/blackjack4494 register
> it, could be useful. I'm also happy to register it on his behalf...

**Comment by FrickTheRIAA** [comment]
I'd highly suggest not including youtube in the domain name. We don't want it to get suspended because of trademark infringement.
Particularly .is domains are extremely hard to take down, pretty much impossible in our case since this project is perfectly legal and the only way to get it taken down is via a court order issued by an Icelandic court.
.ch and .ws are said to be pretty resilient as well in case you want a cheaper option.
No matter what TLD you go with, use a resilient domain registrar/reseller that won't take it down the minute they get a complaint. See the "Resilient domain registrars/resellers" in my [guide](https://old.reddit.com/r/youtubedl/comments/jktx5b/how_to_anonymously_host_the_continued_development/) for more info about that.

**Comment by Nekun** [comment]
@FrickTheRIAA added links to original posts. For any contacts watch uids of my PGP key 55FF381A18B168D3B8C798BDBCF985EA5454B928

**Comment by real-andrewr** [comment]
Hi, I tested out git-bug just out of interest. It's fairly straight forward, took maybe 30 mins to get the hang of it (in terms of setting up a bridge, pushing, pulling).

The way git-bug works is that it stores metadata in the .git directory. I was expecting something else, but this is interesting because you can just store the bugs in the code repository.

There are basically 3 models you could use, if you actually want to use git-bug to offer central bug/issue tracking.

1. Keep issues on GitHub which is effectively authoritative. Users can create issues using git-bug if they want to (you can already do this now). Perhaps someone could run a git-bug cron job and host a backup repository separately, in case GitHub goes down; if this is the case you effectively have to go to (3).

2. Predominantly use git-bug for bug tracking. Users can still raise bugs in GitHub. A cron job would be set up which pulls and pushes bugs between git-bug and GitHub. The advantage is obviously that users can still post bugs and reply in GitHub. The significant downside is that whichever GitHub user account is configured for the git-bug bridge will be the author of any issues and comments. For example, if xloem created an issue in git-bug and the bridge was running from my account, a corresponding issue would be created in GitHub by my account (real-andrewr). I guess you could run the bridge using an intentionally named GitHub account to avoid some confusion, ytdlcbugbot, and basically just offer a degraded experience for GitHub users.

3. Use git-bug to import all issues from GitHub, and then just use git-bug going forward. Downside: users have to figure out how to use git-bug.

P.S. I will put up a git-bug mirror of the issues on yt-dlc anyway, so at least there is a backup of the issues against this new account. Probably will do this in the next day or so.

P.P.S. I'm hosting the mirror https://youtube-dl-sources.org. I'll move it to Gitea in the next day or two (depending on some real life things, like getting my car serviced and work) and put up a little guide on how I set it up, with scripts, cron jobs etc. in case it's useful to anyone. It's nothing too complicated, just takes time to go through all of the doco and actually do it.

**Comment by FrickTheRIAA** [comment]
@Nekun Thanks! However, my old, bad suggestion of running Tor over VPN is still there in your mirror. Could you update the whole guide so that it's a copy of the current version on [Reddit](https://old.reddit.com/r/youtubedl/comments/jktx5b/how_to_anonymously_host_the_continued_development/)? That's what I was trying to say in my previous post, but it might have sounded unclear. Thanks and sorry for the inconvenience!

**Comment by quyleanh** [comment]
Hear something bad on [xda](https://www.xda-developers.com/github-warns-of-bans-for-users-still-distributing-youtube-dl/). Please consider asap @blackjack4494

**Comment by ssokolow** [comment]
@quyleanh That was covered on TorrentFreak two days ago.

> “Please note that **re-posting the exact same content** that was the subject of a takedown notice without following the proper process is a violation of GitHub’s DMCA Policy and Terms of Service. 

(Emphasis mine)

It's about the flurry of people reuploading the same repo that got taken down without adding any new commits to make an effort to remedy the reason for the takedown.

**Comment by w3bb** [comment]
I'm pro-email. It's decentralised and git is designed for it. If we choose any centralised solution we're right back where we started.

**Comment by SoniEx2** [comment]
mailing lists are not, in fact, decentralized. and if you ditch mailing lists you become completely undiscoverable.

**Comment by ssokolow** [comment]
Yeah. What makes mailing lists valuable is that they can be self-hosted (like Gitea) and that they replicate all pieces of data to all clients (like git with something like git-bug to merge in tracked bugs).

**Comment by ohnonot** [comment]
It seems that the "real" youtube-dl is back in business, and off github (on that other git-something).
Unfortunately, bandcamp still doesn't work on 2020.11.01.1 :(
That makes me sad, because I'd much prefer to keep (main) development centralised.
Where is the code base on this one? Will it be able to re-base itself on the "original", again and again, to avoid forking hell?

**Comment by Seirdy** [comment]
@ohnonot youtube-dlc pre-dates the DMCA takedown. It was created to
address inactivity; the original maintainers weren't able to merge PRs
at a good pace.

**Comment by blackjack4494** [comment]
> It seems that the "real" youtube-dl is back in business, and off github (on that other git-something).
> Unfortunately, bandcamp still doesn't work on 2020.11.01.1 :(
> That makes me sad, because I'd much prefer to keep (main) development centralised.
> Where is the code base on this one? Will it be able to re-base itself on the "original", again and again, to avoid forking hell?

initial idea was actually to keep similar/same structure as youtube-dl so you could easily merge changes back to main project. However addressing issues in youtube-dl on github got me banned. I offered them help and so did a few others but they refused as they said there are enough people having write rights.
What forking hell? A fork is for people developing their own solutions and creating PRs to merge back changes. Another case is that you start your own project on that fork and later set it up as a dedicated repository.
As far as I know there is no real active development in any other forks.  
Bandcamp should work with youtube-dlc tho.
